src/protocol/templates/ping1d.cpp.in

```diff
@@ -0,0 +1,213 @@
+/*
+* ping1d.cpp
+* A device API for the Blue Robotics Ping1D echosounder
+*
+*~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!
+* THIS IS AN AUTOGENERATED FILE
+* DO NOT EDIT
+*~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!
+*
+*/
+
+#include "ping1d.h"
+
+Ping1D::Ping1D(Stream& ser, uint32_t baudrate) : _stream ( ser ) {}
+
+Ping1D::~Ping1D()
+{
+{% for msg in messages["ping1D"]["get"] %}
+{% for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+    if (_{{field.name}}) {
+       free(_{{field.name}});
+    }
+{% endif %}
+{% endfor %}
+{% endfor %}
+}
+
+PingMessage* Ping1D::read()
+{
+    while(_stream.available()) {
+        if (_parser.parseByte(_stream.read()) == PingParser::NEW_MESSAGE) {
+          handleMessage(&_parser.rxMsg);
+          return &_parser.rxMsg;
+        }
+    }
+    return nullptr;
+}
+  
+size_t Ping1D::write(uint8_t* data, uint16_t length)
+{
+    return _stream.write(data, length);
+}
+
+bool Ping1D::initialize(uint16_t ping_interval_ms)
+{
+    if(!request(Ping1DNamespace::Device_id)) {
+        return false;
+    }
+
+    if (!request(Ping1DNamespace::Firmware_version)) {
+        return false;
+    }
+
+    if(!request(Ping1DNamespace::Voltage_5)) {
+        return false;
+    }
+
+    if (!request(Ping1DNamespace::Processor_temperature)) {
+        return false;
+    }
+
+    // Configure ping interval
+    if (!set_ping_interval(ping_interval_ms)) {
+        return false;
+    }
+
+    return true;
+}
+  
+PingMessage* Ping1D::waitMessage(enum Ping1DNamespace::msg_ping1D_id id, uint16_t timeout_ms)
+{
+    uint32_t tstart = millis();
+    while (millis() < tstart + timeout_ms) {
+
+        PingMessage* pmsg = read();
+
+        if (!pmsg) {
+            continue;
+        }
+
+        handleMessage(pmsg);
+
+        if (pmsg->message_id() == Ping1DNamespace::Nack) {
+            ping_msg_ping1D_nack nack(*pmsg);
+
+            // TODO remove, this won't ever be true in the current design
+            if (nack.nacked_id() == id) {
+                return nullptr;
+            }
+        }
+
+        if (pmsg->message_id() == id) {
+            return pmsg;
+        }
+    }
+
+    //debug("timed out waiting for id %d", id);
+    return nullptr;
+}
+  
+void Ping1D::handleMessage(PingMessage* pmsg)
+{
+    switch (pmsg->message_id()) {
+{% for msg in messages["ping1D"]["get"] %}
+        case Ping1DNamespace::{{msg|title}}:
+        {
+          ping_msg_ping1D_{{msg}} m(*pmsg);
+{% for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+            if (m.{{field.name}}_length() > _{{field.name}}_length) {
+                if (_{{field.name}}) {
+                    free(_{{field.name}});
+                }
+              _{{field.name}} = ({{get_type_string(field.vector.datatype)}}*)malloc(_{{field.name}}_length * sizeof({{get_type_string(field.vector.datatype)}}));
```

Comment:
How are you going to free this memory ?

Reply:
I free in destructor, and before reallocation (when necessary). Is there a problem?

Reply:
I see it now, ty